### PR TITLE
Add ability to request detailed distance information from fcl

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
This PR expose the ability to request detailed distance information from FCL.  I followed a similar architecture to the collision request and response. I wanted to get this submitted, but there a few remaining items: Update Collision World FCL and correct formatting.

### CODE REVIEW

**Comment by davetcoleman** [comment]
great contribution! looks like clang-format is off

**Comment by Levi-Armstrong** [comment]
@davetcoleman As a result of this PR it implements a request and response structure like the collision checking methods which eliminate the need for several methods which I have marked with the deprecated macro. As a result this creates about 260 warning messages throughout MoveIt, how would you like this to be handled?

**Comment by davetcoleman** [comment]
Is there an issue with changing the collision checking calls throughout the codebase to the new API? This shouldn't break any public-facing APIs, correct?

**Comment by Levi-Armstrong** [comment]
I will update the codebase to use the new code base.

**Comment by davetcoleman** [comment]
I have restarted the build

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 1]
```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
...
+  DistanceResultsData()
+  {
+    clear();
+  }
+
+  /// The minimum distance between two objects. If two objects are in collision, min_distance <= 0.
+  double min_distance;
+
+  /// The nearest points
+  Eigen::Vector3d nearest_points[2];
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: I guess `nearest_points[i]` refers to a point on `link_name[i]`
Rename to `nearest_point` to be consistent with `link_name`.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 2]
```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
...
+
+  /// The object link names
+  std::string link_name[2];
+
+  /// The gradient
+  Eigen::Vector3d gradient;
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Specify the direction of the gradient: `nearest_point[1] - nearest_point[0]` or the other way around?
Is the vector normalized?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 3]
```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
...
+
+  /// Indicates if gradient was calculated.
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Aren't nearest points and gradient coupled? What do I miss?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 4]
```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
...
+  {
+  }
+  virtual ~DistanceResult()
+  {
+  }
+
+  /// Indicates if two objects were in collision
+  bool collision;
+
+  /// The results for the two objects with the minimum distance
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: I think this needs more documentation. What's the difference between `minimum_distance` and `distance`?
What exactly is stored in the `DistanceMap`?

File: moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h [hunk 5]
```diff
@@ -140,6 +140,18 @@ class CollisionWorldDistanceField : public CollisionWorld
     return 0.0;
   }
 
+  virtual void distanceRobot(const DistanceRequest& req, DistanceResult& res,
+                             const CollisionRobot& robot, const robot_state::RobotState& state) const
+  {
+    logWarn("Not implemented");
```
Comment (rhaschke) [inline moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h]: I think this should be `logError`.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 1 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Ok

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 3 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Yes they are coupled

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 4 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: The DistanceMap stores distance results information for every link in active components list.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h [hunk 6]
```diff
@@ -163,16 +164,19 @@ class CollisionRobot
                                    const AllowedCollisionMatrix& acm) const = 0;
 
   /** \brief The distance to self-collision given the robot is at state \e state. */
+  MOVEIT_DEPRECATED
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: Ok

File: moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h (hunk 5 above)
Comment (Levi-Armstrong) [reply moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h]: I agree, this is the way it was originally, but I am happy to update them.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 7]
```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
...
+    , global(true)
+    , active_components_only(NULL)
+    , acm(NULL)
+    , distance_threshold(std::numeric_limits<double>::max())
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+
+  DistanceRequest(bool detailed, bool global, const std::set<const robot_model::LinkModel*>* active_components_only,
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: >  I think, the collision world functionality should be implemented as well ;-)

I believe they have been implemented as part of the PR. Is there something missing?

> You should prepare another commit, removing the now deprecated functions. 
    
So you want me to remove the deprecated function as part of this PR, wont that break others code?

> I guess the use of the old CollisionData structure becomes obsolete then as well?

This should not have any affect on the Collision checking data structure but they are similar.

[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: >> You should prepare another commit, removing the now deprecated functions.

> So you want me to remove the deprecated function as part of this PR, wont that break others code?

No, the new commit should become another PR, that we can apply e.g. to the Lunar branch (or the next release branch). Just, not to forget about cleaning up...

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 3 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Actually my statement is not correct. In the case where there is not a collision the statement is true but in the case of a collision the nearest points are the same for the fcl case so and the normal comes from contact structure. I recently ran into this and I am pushing changes along with your other requested changes.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 8]
```diff
@@ -294,10 +247,16 @@ struct DistanceRequest
...
-  /// Indicate if detailed information should be calculated
-  bool detailed;
+  /// Indicate if nearest point information should be calculated
+  bool enable_nearest_points;
 
-  /// Indicate if the global minimum should be found
+  /// Indicate if a signed distance should be calculated in a collision.
+  bool enable_signed_distance;
+
+  /// Indicate if the global minimum should only be found be found. If this is true
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: typo: "be found be found"

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 9]
```diff
@@ -294,10 +247,16 @@ struct DistanceRequest
...
+  /// it will only try to find the global minimum distance and not store information 
+  /// on a link by link bases. If this is set to false it will store distance information
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: typo: bases -> basis

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 10]
```diff
@@ -326,48 +285,43 @@ struct DistanceResultsData
...
   /// The object link names
-  std::string link_name[2];
-
-  /// The gradient
-  Eigen::Vector3d gradient;
+  std::string link_names[2];
 
-  /// Indicates if gradient was calculated.
-  bool hasGradient;
+  /// A normalized vector pointing from link_names[0] to link_names[1].
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Shouldn't that read:
> A normalized vector pointing from nearest_points[0] to nearest_points[1]

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 11]
```diff
@@ -224,60 +224,13 @@ struct CollisionRequest
...
     , global(true)
     , active_components_only(NULL)
     , acm(NULL)
     , distance_threshold(std::numeric_limits<double>::max())
     , verbose(false)
-    , gradient(false)
-  {
-  }
-
-  DistanceRequest(bool detailed, bool global, const std::set<const robot_model::LinkModel*>* active_components_only,
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: That's a nice cleanup! +1

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 12]
```diff
@@ -592,53 +592,90 @@ bool distanceDetailedCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2
...
-    dist_result.link_name[1] = cd2->ptr.obj->id_;
+    dist_result.link_names[0] = cd1->ptr.obj->id_;
+    dist_result.link_names[1] = cd2->ptr.obj->id_;
+    dist_result.hasNearestPoints = cdata->req->enable_nearest_points;
+    if (dist_result.hasNearestPoints)
+    {
+      dist_result.normal = (dist_result.nearest_points[1] - dist_result.nearest_points[0]).normalized();
+    }
+
+    if (d <= 0 && cdata->req->enable_signed_distance)
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: Why do you need this extra code block? Shouldn't `fcl::distance` compute the correct distance?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: Not yet addressed.

File: moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h (hunk 5 above)
Comment (rhaschke) [reply moveit_experimental/collision_distance_field/include/moveit/collision_distance_field/collision_world_distance_field.h]: Not yet addressed.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 10 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Yep

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 12 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: That is what I had hoped, but in fcl 0.5 if two objects are in collision it returns 0 or -1 depending on the object type. Now fcl has been making a lot of improvements over the past year and in version 0.6 which has not been release will return this information directly. This interim fix to get this information until the 0.6 release. I also have another branch which I will push this week that make the necessary change for 0.6 since they reorganized everything and switch to using Eigen instead of custom matrix types.

**Comment by rhaschke** [comment]
Could you, please, also rebase your branch onto the latest kinetic-devel head?

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 12 above)
Comment (rhaschke) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: I see.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 10 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Sorry, I believe it should be link_names[0] to link_names[1] mainly because in the event of a collision the nearest_points are the same but the normal is always calculate such that it points from link_names[0] to link_names[1].

**Comment by Levi-Armstrong** [comment]
@rhaschke I believe I have addressed the requested changes.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 13]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+    , enable_signed_distance(false)
+    , global(true)
+    , active_components_only(nullptr)
+    , acm(nullptr)
+    , distance_threshold(std::numeric_limits<double>::max())
+    , verbose(false)
+  {
+  }
+
+  virtual ~DistanceRequest()
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: These structs don't need virtual destructors either.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 14]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+  bool enable_nearest_points;
+
+  /// Indicate if a signed distance should be calculated in a collision.
+  bool enable_signed_distance;
+
+  /// Indicate if the global minimum should only be found. If this is true
+  /// it will only try to find the global minimum distance and not store information
+  /// on a link by link basis. If this is set to false it will store distance information
+  /// for every link in the active_components_only list.
+  bool global;
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: `global` is somewhat generic to read. What about `global_minimum_only` or reversing the logic and using `detailed_distances` or `linkwise_distances`?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 15]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+
+  std::string group_name;
+
+  /// The set of active components to check
+  const std::set<const robot_model::LinkModel*>* active_components_only;
+
+  /// The allowed collision matrix used to filter checks
+  const AllowedCollisionMatrix* acm;
+
+  /// A distance threshold to reduce number of queries
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: What's the exact meaning of this variable? Only consider distances / collisions above or below this value?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 16]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+  double distance_threshold;
+
+  /// Log debug information
+  bool verbose;
+
+  /// Indicate if gradient should be calculated between each object. This the vector defined by the line connecting the
+  /// closest points on the two objects.
+  bool gradient;
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Rename to `compute_gradient`.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 17]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+  /// The nearest points
+  Eigen::Vector3d nearest_points[2];
+
+  /// The object link names
+  std::string link_names[2];
+
+  /// A normalized vector pointing from link_names[0] to link_names[1].
+  Eigen::Vector3d normal;
+
+  /// Indicates if nearest points were found.
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: That is, whether `nearest_points[]` is valid?
What are reasons, that nearest point are not found?
Please augment the documentation here.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h [hunk 18]
```diff
@@ -163,27 +163,81 @@ class CollisionRobot
                                    const AllowedCollisionMatrix& acm) const = 0;
 
   /** \brief The distance to self-collision given the robot is at state \e state. */
-  virtual double distanceSelf(const robot_state::RobotState& state) const = 0;
+  inline double distanceSelf(const robot_state::RobotState& state) const
```
Comment (rhaschke) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: All these old functions (that you deprecated) are now implemented as thin inline wrappers of the actual worker function. Please verify correctness.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 19]
```diff
@@ -404,29 +405,33 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
+      active2 = false;
+
     // If neither of the involved components is active
-    if ((!l1 || cdata->active_components_only_->find(l1) == cdata->active_components_only_->end()) &&
-        (!l2 || cdata->active_components_only_->find(l2) == cdata->active_components_only_->end()))
+    if (active1 == false && active2 == false)
     {
-      min_dist = cdata->res_->distance;
-      return cdata->done_;
+      return false;
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: Why did you dropped setting `min_dist`? Probably, that's fine. I just don't understand the relevance of the previous code. I'm not even sure that the right-hand value had any valid initialization before.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 20]
```diff
@@ -404,29 +405,33 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
-    bool found = cdata->acm_->getAllowedCollision(cd1->getID(), cd2->getID(), type);
+    bool found = cdata->req->acm->getAllowedCollision(cd1->getID(), cd2->getID(), type);
     if (found)
     {
       // if we have an entry in the collision matrix, we read it
       if (type == AllowedCollision::ALWAYS)
       {
         always_allow_collision = true;
-        if (cdata->req_->verbose)
+        if (!cdata->req->verbose)
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: Why did you reversed the meaning of `verbose` everywhere in this code???

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 21]
```diff
@@ -463,37 +468,148 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
 
   if (always_allow_collision)
   {
-    min_dist = cdata->res_->distance;
-    return cdata->done_;
+    return false;
   }
 
-  fcl::DistanceResult dist_result;
-  dist_result.update(cdata->res_->distance, nullptr, nullptr, fcl::DistanceResult::NONE,
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: You dropped this `update` call. Can you elaborate on this?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 15 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: I will update the documentation, but it will only calculated distance for objects within this threshold to each other.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 17 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: I wrote most of this a few years ago so not sure what my intent was at the time, but after reviewing it it should probably be removed.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h (hunk 18 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: I am not sure what has happened but on my pc these changes are not present.

Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: It looks like this was a change you made in the commit "keep old functions as conviencency wrappers". It is interesting that you were able to change code in my repository.

I will let you make this correction then I will apply the other requested fixes.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 19 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: The min_dist does the same thing as distance_threshold in the distance request when searching for global minimum. Now that we enable the ability to get not only the global minimum but the ability to get information about every link, the information is now stored on a link by link bases in the distance request.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 20 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: This was done in error. I will correct it.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 21 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: This was a change you made as part of the commit "replace distanceCallback with new (detailed) function".

Comment (rhaschke) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: According to git blame, yes, these changes are done by me. However, I only cleaned up your changes in order to facilitate the review. However, this update function call was already missing in your `distanceDetailedCallback()` function. Only with my cleanup changes, I was able to notice this change... Hence, the question is, why was it removed in `distanceDetailedCallback`?

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h (hunk 18 above)
Comment (rhaschke) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: I was done with my changes. You can proceed ;-)

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 22]
```diff
@@ -217,6 +220,139 @@ struct CollisionRequest
...
+
+  /// The distance between two objects. If two objects are in collision, distance <= 0.
+  double distance;
+
+  /// The nearest points
+  Eigen::Vector3d nearest_points[2];
+
+  /// The object link names
+  std::string link_names[2];
+
```
Comment (BryceStevenWilley) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Might be worth it to add `BodyType types[2];`, so you can easily tell which body(s) are from the robot or the environment.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 16 above)
Comment (rhaschke) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: The `gradient` variable is not initialized in the constructor!

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h (hunk 18 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_robot.h]: @rhaschke Was it your intention to move these to the base class?

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 21 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: Oh, after the object is constructed the only variable that actually needed to be changed is the min_distance which used by fcl to only do detailed distance calculations if two objects are within this distance. The update call was unnecessary since the object was just constructed.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 22 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: @rhaschke All requested changes should be addressed now.

**Comment by Levi-Armstrong** [comment]
@rhaschke All requested changes should be addressed now.

File: moveit_ros/benchmarks/src/BenchmarkExecutor.cpp [hunk 23]
```diff
@@ -850,6 +850,7 @@ void BenchmarkExecutor::collectMetrics(PlannerRunData& metrics,
           correct = false;
         if (!p.getWayPoint(k).satisfiesBounds())
           correct = false;
+
```
Comment (davetcoleman) [inline moveit_ros/benchmarks/src/BenchmarkExecutor.cpp]: this change does not belong in this PR

Comment (Levi-Armstrong) [reply moveit_ros/benchmarks/src/BenchmarkExecutor.cpp]: I will correct it.

**Comment by Levi-Armstrong** [comment]
@davetcoleman and @rhaschke, I am currently making one more change to improve the storage of distance information similar to collision contacts. In some cases (Optimization Motion Planning) it would be useful to have all objects with a distance to a given link instead of just the closest one.

**Comment by Levi-Armstrong** [comment]
@davetcoleman and @rhaschke  I have made the change let me know if you would like anything changed.

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h [hunk 24]
```diff
@@ -221,12 +221,21 @@ struct CollisionRequest
   bool verbose;
 };
 
+enum DistanceRequestType
```
Comment (davetcoleman) [inline moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: Namespaceify your enums, e.g.:   ``namespace DistanceRequestType {``

"This prevents enums from polluting the namespace they're inside. Individual items within the enum are referenced by: Choices::Choice1, but the typedef still allows declaration of the Choice enum without the namespace."
http://wiki.ros.org/CppStyleGuide

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 25]
```diff
@@ -550,62 +534,63 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
 
-    if (!cdata->req->global_minimum_only)
+    if (cdata->req->type != DistanceRequestType::GLOBAL)
     {
-      if (d <= 0 && !cdata->res->collision)
+      if (it == cdata->res->distances.end())
       {
-        cdata->res->collision = true;
+        std::vector<DistanceResultsData> data;
+        data.reserve(cdata->req->max_contacts_per_body);
```
Comment (davetcoleman) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: i don't understand why you reserve this vector then only ``push_back`` one item. i don't see a for loop anywhere else to fill this in.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 26]
```diff
@@ -474,45 +468,37 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
-    it2 = cdata->res->distances.find(cd2->ptr.obj->id_);
+  const std::pair<std::string, std::string>& pc = cd1->getID() < cd2->getID() ?
+                                                        std::make_pair(cd1->getID(), cd2->getID()) :
+                                                        std::make_pair(cd2->getID(), cd1->getID());
+
+  DistanceMap::iterator it = cdata->res->distances.find(pc);
 
-    if (cdata->req->active_components_only)
+
+  if (it != cdata->res->distances.end())
```
Comment (davetcoleman) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: all this logic you've modified could really use some explanation comments

File: moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h (hunk 24 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h]: I will fix it.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 25 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: The loop is hidden within fcl, it calls the callback function multiple times.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 26 above)
Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: Before the latest commit it was only possible to get the global minimum distance or the minimum distance for a given link. This logic change adds the ability to get all of the contacts for a given pair of links. This is useful if a link is composed of multiple geometries and by getting all of the contacts you can accurately define a gradient for those pair of links to move them out of collision.

**Comment by Levi-Armstrong** [comment]
@davetcoleman and @rhaschke I have made the requested changes.

File: moveit_core/collision_detection_fcl/src/collision_common.cpp (hunk 25 above)
Comment (davetcoleman) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: please add this as a comment since its not clear why ``reserve`` is otherwise needed and it would run fine without it (its a hidden optimization)

**Comment by rhaschke** [comment]
This is on my todo list. But there is a more urgent deadline this week. Please be patient some more days.

**Comment by davetcoleman** [comment]
@rhaschke its been 5 months, I think we should really move forward with this so other PRs like https://github.com/ros-planning/moveit/pull/746 don't break this one

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 27]
```diff
@@ -463,37 +462,141 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
-    cdata->res_->distance = -1;
+    if (cdata->req->type == DistanceRequestType::LIMITED)
+    {
+      // If at the limit for a given pair just return
+      if (it->second.size() >= cdata->req->max_contacts_per_body)
+      {
+        return cdata->done;
+      }
+    }
+    else if (cdata->req->type != DistanceRequestType::GLOBAL)
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: These if-checks do not make sense to me. Shouldn't they equality instead of inequality checks?

File: moveit_core/collision_detection_fcl/src/collision_common.cpp [hunk 28]
```diff
@@ -463,37 +462,141 @@ bool distanceCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2, void*
...
+            max_index = i;
+          }
+        }
 
-  return cdata->done_;
+        const fcl::Contact& contact = coll_res.getContact(max_index);
+        dist_result.distance = -contact.penetration_depth;
+        dist_result.nearest_points[0] = Eigen::Vector3d(contact.pos.data.vs);
+        dist_result.nearest_points[1] = Eigen::Vector3d(contact.pos.data.vs);
+        dist_result.normal = Eigen::Vector3d(contact.normal.data.vs);
```
Comment (rhaschke) [inline moveit_core/collision_detection_fcl/src/collision_common.cpp]: Here you deviate from your definition that the normal vector connects nearest points. This should be documented.

Comment (Levi-Armstrong) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: This is a deviation. If you are using convex to convex checking and in contact it would be as documented but in the case of non-convex to non-convex fcl provides the contact normal (One of the triangle normals). From what I could tell the way it choose which triangle normal to use it picked the one with the least penetration. Also for non-convex to non-convex if the objects are not in contact it is as documented. I something along what I just described sufficient to add?

In the case of moveit there is not a way to tell fcl to use convex representation so currently everything uses non-convex but it would be nice to support this in the future.

Comment (rhaschke) [reply moveit_core/collision_detection_fcl/src/collision_common.cpp]: Yes, such an explanation will be fine.

**Comment by Levi-Armstrong** [comment]
@davetcoleman and @rhaschke the remaining two requests should be complete.

**Comment by davetcoleman** [comment]
There's a minor clang-format issue failing Travis

**Comment by davetcoleman** [comment]
Hurray!!  :clap:

@Levi-Armstrong want to announce this change on the Discourse channel?

**Comment by BoltzmannBrain** [comment]
Thanks for doing this @Levi-Armstrong! Just caught up here, from [this ROS-answers thread](https://answers.ros.org/question/252110/distancetocollision-within-a-moveit-planner/) where I asked about a collision distance per robot link. How can I now implement this feature? Currently I have something like,
```
auto req = collision_detection::CollisionRequest();
req.distance = true;  // Compute single proximity distance value
```

**Comment by BoltzmannBrain** [comment]
I've been using the new functionality in one of my feature projects, working off the tip of kinetic-devel. For our workflow, however, the rest of the system uses the most recent release (i.e. 0.9.11). When can we expect a new release to be cut (that includes this merged PR)? @davetcoleman @rhaschke

**Comment by davetcoleman** [comment]
Good question, please move this discussion to https://github.com/ros-planning/moveit/issues/886
